src/ImageSharp/IO/ChunkedMemoryStream.cs

```diff
@@ -0,0 +1,544 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+using SixLabors.ImageSharp.Memory;
+
+namespace SixLabors.ImageSharp.IO
+{
+    /// <summary>
+    /// Provides an in-memory stream composed of non-contiguous chunks that doesn't need to be resized.
+    /// Chunks are allocated by the <see cref="MemoryAllocator"/> assigned via the constructor
+    /// and is designed to take advantage of buffer pooling when available.
+    /// </summary>
+    internal sealed class ChunkedMemoryStream : Stream
+    {
+        /// <summary>
+        /// The default length in bytes of each buffer chunk.
+        /// </summary>
+        public const int DefaultBufferLength = 4096;
+
+        // The memory allocator.
+        private readonly MemoryAllocator allocator;
+
+        // Data
+        private MemoryChunk memoryChunk;
+
+        // The length of each buffer chunk
+        private readonly int chunkLength;
+
+        // Has the stream been disposed.
+        private bool isDisposed;
+
+        // Current chunk to write to
+        private MemoryChunk writeChunk;
+
+        // Offset into chunk to write to
+        private int writeOffset;
+
+        // Current chunk to read from
+        private MemoryChunk readChunk;
+
+        // Offset into chunk to read from
+        private int readOffset;
```

Comment:
I'm probably missing something, but why are we writing/reading to/from different chunks?
Shouldn't `Stream.Position` indicate where we are in both cases?

Reply:
For my sanity mostly. This is hard.

Reply:
Ah ok, so it's a special stream implementation where writes go always to the end of the stream, right? Would add a comment about this.